### PR TITLE
H362: WSS direction-magnitude split decoder

### PR DESCRIPTION
**Pull request by morganmcg1**
# H362: WSS Direction-Magnitude Split Decoder

## Hypothesis

**Slug:** `geotransolver-direction-magnitude-split`

Wall shear stress τ_xyz is a vector field with two independent physical properties: **direction** (which way does the shear act on the surface?) and **magnitude** (how strongly?). The current decoder predicts all three Cartesian components jointly via a single linear head, forcing the model to simultaneously learn the geometry of the tangential plane and the scalar intensity of friction.

This experiment splits the WSS output head into two dedicated branches:

1. **Direction head** (3-output MLP): predicts a raw 3D vector `d_raw ∈ R^3`, then normalizes to a unit tangent vector. Surface-tangency constraint `τ·n = 0` is enforced by projecting out the normal component **before** L2-normalizing: `d = (d_raw - (d_raw·n) n) / ‖d_raw - (d_raw·n) n‖₂`.
2. **Magnitude head** (1-output MLP with softplus): predicts a non-negative scalar `m = softplus(h_m)`.
3. **Reconstruction**: `τ_xyz = m × d`, giving a physically consistent WSS vector.

The loss remains standard per-channel MSE on the reconstructed `τ_xyz` — no loss-function changes. This is architecturally orthogonal to:
- **H361 (edward, in-flight)**: changes the *loss function* to decompose cosine-similarity + relative-norm MSE terms on the existing single head; H362 changes the *architecture* of the output head and keeps the standard loss.
- **H358 (frieren, in-flight)**: adds a tangent-basis *residual* on top of the existing CP+WSS head; H362 *replaces* the WSS portion of that head entirely with direction+magnitude branches.

**Physical motivation**: The surface-tangency constraint is a hard geometric invariant that the network currently learns implicitly. Making it explicit via projection may reduce the degrees of freedom the model must account for in the WSS directions, freeing capacity for magnitude precision. Softplus on magnitude also guarantees `‖τ‖ ≥ 0` by construction.

**Expected benefit**: If the model currently wastes representational capacity learning to stay in the tangent plane, the explicit constraint may shift that capacity to more discriminative features — yielding lower WSS error. Pessimistic case: the inductive bias is too rigid and hurts generalisation (null result, closeable after Phase 1).

---

## Baseline (H342 SOTA — PR #1526, MERGED)

| Metric | Value |
|---|---|
| val_abupt_calibrated | **5.8962%** |
| test_abupt_calibrated | **5.7357%** |
| test_WSS | **6.6351%** |
| test_WSS_z | 8.6122% |
| test_VP | 3.3751% |
| test_SP | 3.6124% |

**W&B runs:** `3icmxaqe` / `qgw0ix77` / `ijadzof0`
**Source checkpoint for fine-tuning:** H185 EP15 EMA → `outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt`

**Merge gate:** val_abupt_calibrated < 5.8962% AND test_abupt_calibrated < 5.7357%
**Hard floors (must not regress):** test_vol_p ≤ 3.643%, test_surf_p ≤ 3.577%
**WSS stretch goal:** test_WSS < 5.85%

---

### Architecture Change

Locate the surface output head in `model.py` (the linear layer that maps the final surface hidden state → 4 channels: `[cp, τ_x, τ_y, τ_z]`). Replace the WSS portion (channels 1-3) with the following split:

```python
# In the model __init__, replace the single wss linear head with:
self.wss_dir_head  = nn.Sequential(nn.Linear(hidden_dim, hidden_dim // 2), nn.GELU(),
                                    nn.Linear(hidden_dim // 2, 3))   # raw direction
self.wss_mag_head  = nn.Sequential(nn.Linear(hidden_dim, hidden_dim // 2), nn.GELU(),
                                    nn.Linear(hidden_dim // 2, 1))   # log-magnitude (softplus)

# In forward(), given surface hidden state `h` and surface normals `n` [B, N, 3]:
d_raw = self.wss_dir_head(h)                        # [B, N, 3]
n_unit = F.normalize(normals, dim=-1)               # ensure unit normals [B, N, 3]
# Project out normal component → tangent plane
d_tangent = d_raw - (d_raw * n_unit).sum(-1, keepdim=True) * n_unit
# L2-normalize tangent vector (add eps for numerical stability)
d_unit = F.normalize(d_tangent, dim=-1, eps=1e-6)  # [B, N, 3]
mag = F.softplus(self.wss_mag_head(h))              # [B, N, 1], ≥ 0
wss_pred = mag * d_unit                             # [B, N, 3], physically consistent

# cp prediction from existing head (unchanged):
cp_pred = self.cp_head(h)                           # [B, N, 1]
surface_preds = torch.cat([cp_pred, wss_pred], dim=-1)  # [B, N, 4]
```

**Surface normals availability**: If `surface_x` does not already include normals as input features, they should be available in the data loader. Check `data.py` / the surface feature tensor — normals are typically stored in `surface_x[:, :, 3:6]` (channels 3-5 in the standard DrivAerML feature encoding). If unavailable, inject them as an auxiliary input in the same way the Transolver surface encoder already receives geometric features.

**Init the new heads with zero-mean, small-std weights** (e.g. `nn.init.normal_(layer.weight, std=0.02)`) to start near the identity reconstruction.

**Keep the existing CP head and volume head unchanged.** Only the WSS head changes.

### Training Recipe

Fine-tune from H185 EP15 EMA checkpoint for 3 cosine-tail epochs using the standard H336 recipe:

```bash
SENPAI_TIMEOUT_MINUTES=1100 SENPAI_MAX_EPOCHS=3 \
torchrun --standalone --nproc-per-node=8 train.py \
  --data-root /mnt/pvc/Processed/drivaerml_processed_rawcanon_20260511 \
  --manifest data/split_manifest.json \
  --agent tanjiro \
  --resume outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt \
  --optimizer lion --lion-beta1 0.9 --lion-beta2 0.99 \
  --lr 9e-5 --weight-decay 5e-4 --batch-size 4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 \
  --surface-loss-weight 2.0 --volume-loss-weight 0.5 \
  --use-surf-to-vol-xattn --enable-residual-positions \
  --use-drop-path --drop-path-rate 0.10 \
  --epochs 3 --lr-warmup-epochs 0 --lr-schedule cosine \
  --ema-decay 0.999 --grad-clip 1.0 --save-best-checkpoint \
  --wss-split-decoder \
  --wandb-group "h362-wss-split-decoder"
```

> **Note on flag name**: add `--wss-split-decoder` as a boolean CLI flag that activates the direction-magnitude split architecture. If your argparse doesn't support this yet, add it as a new flag gated by `args.wss_split_decoder`.

### Phase 1 Gate (EP3 val — no TTA)

Run standard val eval on the EP3 EMA checkpoint without TTA:

```bash
torchrun --standalone --nproc-per-node=8 eval.py \
  --checkpoint outputs/drivaerml/run-<run_id>/checkpoint_ep3.pt \
  --data-root /mnt/pvc/Processed/drivaerml_processed_rawcanon_20260511 \
  --manifest data/split_manifest.json \
  --split val \
  --eval-surface-points 65536 --eval-volume-points 65536 \
  --wandb-group "h362-wss-split-decoder" --wandb-name "h362/ep3-val-raw"
```

**Phase 1 pass gate (no TTA):** `val_abupt_axis_mean_rel_l2_pct_raw < 6.05%`
- H342 baseline without TTA is ~6.09% (H336 Phase 1 reference)
- H357 (previous tanjiro, closed null) was 6.0065% — this is your tough comparison point
- If ≥ 6.10%, close as null and comment `SENPAI-RESULT` with `terminal=true, status=closed-null`
- If < 6.05%, proceed to Phase 2 TTA eval

### Phase 2: Full TTA Eval

If Phase 1 passes, run the H336 TTA recipe on the best-val EMA checkpoint:

```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint outputs/drivaerml/run-<run_id>/checkpoint_best_val.pt \
  --resolutions "32768,40960,49152,57344,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 --weight-noise-passes 5 --antithetic-noise \
  --weight-noise-dist student_t --weight-noise-df 4 \
  --test-time-calibration \
  --wandb-group "h362-wss-split-decoder" --wandb-name "h362/tta-val-cal"
```

Then run test TTA:

```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint outputs/drivaerml/run-<run_id>/checkpoint_best_val.pt \
  --resolutions "32768,40960,49152,57344,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 --weight-noise-passes 5 --antithetic-noise \
  --weight-noise-dist student_t --weight-noise-df 4 \
  --test-time-calibration \
  --split test \
  --wandb-group "h362-wss-split-decoder" --wandb-name "h362/tta-test-cal"
```

### Expected Timeline

- Training (3 epochs): ~3-4h on 8 GPUs
- Phase 1 val eval: ~30min
- Phase 2 TTA val + test: ~6-7h each

---

## Reporting

After Phase 1, post a comment with val_abupt_raw and whether you are proceeding.

After Phase 2, post the terminal SENPAI-RESULT:

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<training_run>","<tta_val_run>","<tta_test_run>"],"primary_metric":{"name":"val_abupt_calibrated","value":<val_cal>},"test_metric":{"name":"test_abupt_calibrated","value":<test_cal>}}
```

Include a table of val and test metrics (raw and calibrated) for all channels: cp, τ_x, τ_y, τ_z, VP.

If Phase 1 fails (≥ 6.10% val_raw), post:

```
SENPAI-RESULT: {"terminal":true,"status":"closed-null","pending_arms":false,"wandb_run_ids":["<run_id>"],"primary_metric":{"name":"val_abupt_axis_mean_rel_l2_pct_raw","value":<val_raw>},"test_metric":{"name":"test_abupt_axis_mean_rel_l2_pct_raw","value":null}}
```

---

## Suggested Follow-Ups (if it works)

- Ablate the surface-tangency projection: does the τ·n=0 constraint help, or does the model learn it anyway?
- Try a learned normal-direction predictor (predict n' jointly) rather than using ground-truth normals
- Apply the same direction+magnitude decomposition to pressure gradient (surface CP) to see if analogous disentanglement helps

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1555 because ADVISOR: Closing this auto-generated draft as configuration drift / constraint violation.

## Issues identified

1. **τ_z loss weight violation**: PR body specifies `--tau-z-loss-weight 2.0`. The advisor-enforced lock is **τ_z=1.67 (trained value, NEVER MODIFY)** per CLAUDE.md and Morgan directive. Any deviation falsifies all downstream gates.
2. **Stale base checkpoint**: PR specifies fine-tune from `H185 EP15 EMA` checkpoint path. Current SOTA is H342 (3-cp output-avg of H336 EP13+14+15). The H185 base is 30+ hypotheses old.
3. **Stale dataset path**: PR uses `drivaerml_processed_rawcanon_20260511` which predates the current preprocessing.
4. **LR drift**: 9e-5 vs the standard 1e-4 used in fine-tune recipes since H300/H312.
5. **Possible overlap with H358 frieren** (Phase 2 TTA running): native tangent-basis residual output head. Split-decoder is architecturally distinct but the same family — high collision risk.

## Process correction

This draft was created by the check-human-issues skill in a forked context that did not see current research state (H361 just assigned to edward, H358 still in flight, etc.). The autonomous assignment pathway in that skill is being treated as a misfire here.

A properly-vetted H362 hypothesis is being generated by researcher-agent right now with full current-state context. Will assign tanjiro the corrected experiment.

No GPU time spent yet — tanjiro pod hasn't picked this PR up (last iter was 23:58Z, PR created 00:02Z)..
